### PR TITLE
sql: complex range functions, pt 0

### PR DESCRIPTION
**Pull request by sploiselle**
Implements a set of binary functions for range types.

Note that there are no docs for the `range` type yet (#17340), so not including a release note is intentional.

### Motivation

This PR adds a known-desirable feature.

### Tips for reviewer

@mjibson I added PG tests here because these tests' output is hard for a human to parse; this is meant to ensure that our implementation lines up with PG but that seems like an uncommon/unexpected use of these tests. Is that something you're OK with?

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by sploiselle** [comment]
@MaterializeInc/surfaces PTAL

File: src/expr/src/scalar/func.rs [hunk 1]
```diff
@@ -2878,6 +2935,15 @@ impl fmt::Display for BinaryFunc {
...
             }
+            BinaryFunc::RangeContainsRange { rev, .. } => {
+                f.write_str(if *rev { "<@" } else { "@>" })
+            }
+            BinaryFunc::RangeOverlaps => f.write_str("&&"),
+            BinaryFunc::RangeAfter => f.write_str(">>"),
+            BinaryFunc::RangeBefore => f.write_str("<<"),
+            BinaryFunc::RangeOverleft => f.write_str("&<"),
+            BinaryFunc::RangeOverright => f.write_str("&>"),
+            BinaryFunc::RangeAdjacent => f.write_str("-|-"),
```
Comment (mjibson) [inline src/expr/src/scalar/func.rs:2946]: is this an operator we parse?

**Review by mjibson — APPROVED** [review]
LGTM assuming pgtest and slt results were all checked against postgres.

**Review by jkosh44 — APPROVED** [review]

File: src/expr/src/scalar/func.rs (hunk 1 above)
Comment (sploiselle) [reply src/expr/src/scalar/func.rs:2946]: Yep https://github.com/MaterializeInc/materialize/pull/17341/files#diff-628ea3914035abe87302037eb9284f9a6478f190dfb57ac0f54310c92e14f51fR837

**Comment by sploiselle** [comment]
@mjibson Does `pgtest` automatically check its output against Postgres? I validated most but not all of the functions and was hoping `pgtest` could do some of that for me.

**Comment by mjibson** [comment]
> 

It does: see

**Comment by mjibson** [comment]
> @mjibson Does `pgtest` automatically check its output against Postgres? I validated most but not all of the functions and was hoping `pgtest` could do some of that for me.

It does! See [the docs](https://dev.materialize.com/api/rust-private/mz_pgtest/index.html#usage-while-writing-tests):

> Then run the pgtest binary, enabling rewrites and pointing it at postgres:
> REWRITE=1 cargo run --bin mz-pgtest -- test/pgtest/test.pt --addr localhost:5432 --user postgres

**Comment by sploiselle** [comment]
@mjibson Yep--that was what I thought. I must have just misread your comment: 

> assuming pgtest and slt results were all checked against postgres

File: src/expr/src/scalar.proto [hunk 2]
```diff
@@ -588,7 +588,14 @@ message ProtoBinaryFunc {
...
         google.protobuf.Empty mod_uint64 = 171;
-        ProtoRangeContainsElemInner range_contains_elem = 172;
+        ProtoRangeContainsInner range_contains_elem = 172;
+        bool range_contains_range = 173;
+        google.protobuf.Empty range_overlaps = 174;
+        google.protobuf.Empty range_after = 175;
+        google.protobuf.Empty range_before = 176;
+        google.protobuf.Empty range_overleft = 177;
+        google.protobuf.Empty range_overright = 178;
+        google.protobuf.Empty range_adjacent = 179;
```
Comment (aalexandrov) [inline src/expr/src/scalar.proto:598]: nit: the `Arbitrary for BinaryFunc` is missing these variants (and probably some more), which means that we won't catch possible protobuf errors for those variants.

**Pull request merged**